### PR TITLE
optimize copy times and remove dai device from copier

### PR DESCRIPTION
**Pull request by btian1**
After SOF upgraded to IPC4, copier actually created 2 devices for
data input(capture) and output(playback), it is
 dai and copier,
hence there are two times copy from DMA buffer
to endpoint buffer then to component buffer.

This PR intend to remove dai device and one more time copy
through use dai data to handle host interface calling.

Basically, there are several steps to go there:

Add dai data to copier data structure.
Share dai data structure to copier by add a dai copier common header file.
In copier, all dai related ops calling should be replaced with new dai exposed interface.
Remove one more time copy and dai endpoint buffer.
Remove dai device.
Performance data:

mcps, saved 1.6mcps, 5.4 --> 3.8, around 30% saved for copier.
Before: 0x40001 perf comp_copy samples 48 period 1000 cpu avg 533 peak 553
After:   0x40001 perf comp_copy samples 48 period 1000 cpu avg 378 peak 397

memory, since host endpoint buffer and host device got removed, for each copier,
16bit/32bit 384/768 bytes saved for buffer, 112 saved for dai device.
As a summary, 16/32 bit playback/capture totally saved 496/884 bytes for each dai copier usage.
For the cases that have more than 10 copier usage, memory saving is quite remarkable.

Note：multiple sink dai copier is not tested and will be fixed in follow up PR.

Signed-off-by: Baofeng Tian baofeng.tian@intel.com

### CODE REVIEW

File: src/audio/copier/copier.c
Comment (RanderWang) [inline src/audio/copier/copier.c:746]: we can remove "dev->ipc_config.type > SOF_COMP_DAI", let it like  " if (cd->ipc_gtw) "

Comment (btian1) [reply src/audio/copier/copier.c:746]: I also want to remove, however, there is an internal CI error, I still don't know the reason, for prepare and trigger, if use driver to call, there is no issue, if use patch with current implementation, there is a issue.

This is the reason I still don't remove DAI part yet, just in case.

**Comment by gkbldcig** [comment]
Can one of the admins verify this patch?

**Comment by btian1** [comment]
> Can one of the admins verify this patch?

this is still under development, please stay tuned.

File: src/ipc/ipc4/dai.c
Comment (lgirdwood) [inline src/ipc/ipc4/dai.c]: I think we can probably make some of these light function static inline in the header.

File: src/audio/copier/copier.c
Comment (lgirdwood) [inline src/audio/copier/copier.c:1547]: Needs some inline commenst to explain what this is doing.

File: src/audio/dai-zephyr.c
Comment (lgirdwood) [inline src/audio/dai-zephyr.c:1279]: this could be an inline in the header.

**Review by lgirdwood — COMMENTED** [review]
@juimonen @abonislawski @singalsu pls review

**Comment by lgirdwood** [comment]
> mcps, saved 1.6mcps, 5.4 --> 3.8, around 30% saved for copier.
> Before: 0x40001 perf comp_copy samples 48 period 1000 cpu avg 533 peak 553
> After: 0x40001 perf comp_copy samples 48 period 1000 cpu avg 378 peak 397

@btian1 this is a good saving for link copier. Good work.

**Comment by btian1** [comment]
> pier. Good work.

this is with log print data, actual saving for MCPS should be larger, due to zephyr log print always have impact on performance measure.

File: src/audio/dai-zephyr.c
Comment (RanderWang) [inline src/audio/dai-zephyr.c:304]: return -ENODEV;  error label does nothing now

File: src/audio/dai-zephyr.c
Comment (RanderWang) [inline src/audio/dai-zephyr.c:340]: comp_err(dev, ?

**Review by RanderWang — COMMENTED** [review]
why do we need to create endpoint buffer for dai ? when copier is bound to another module, the buffer is created. Dai can reuse copier buffer. please check host part.

File: src/audio/copier/copier.c
Comment (lyakh) [inline src/audio/copier/copier.c:741]: `else if`. This time `cd->ipc_gtw` is unimportant? Or you could use `switch (dev->ipc_config.type)`

File: src/audio/copier/copier.c
Comment (lyakh) [inline src/audio/copier/copier.c:902]: `switch()`

File: src/audio/copier/copier.c
Comment (lyakh) [inline src/audio/copier/copier.c:978]: `switch ()`

File: src/audio/copier/copier.c
Comment (lyakh) [inline src/audio/copier/copier.c:1027]: `switch ()`

File: src/audio/dai-zephyr.c
Comment (lyakh) [inline src/audio/dai-zephyr.c:1057]: superfluous

File: src/audio/copier/copier.c
Comment (lyakh) [inline src/audio/copier/copier.c:1779]: space missing

File: src/audio/copier/copier.c
Comment (lyakh) [inline src/audio/copier/copier.c:1799]: the indentation was correct

File: src/audio/copier/copier.c
Comment (lyakh) [inline src/audio/copier/copier.c:1801]: actually both these lines aren't needed

File: src/audio/copier/copier.c
Comment (btian1) [reply src/audio/copier/copier.c:728]: config->type = SOF_COMP_HOST;

gtw assigned type as HOST, so for DAI case, there is no possible with ipc_gtw == true.

File: src/audio/dai-zephyr.c
Comment (serhiy-katsyuba-intel) [inline src/audio/dai-zephyr.c:316]: It looks like this "else" processing here is unnecessary. In case of capture, dev->bsink_list should not be empty but should contain one or more regular sink buffers. That case is already handled below in a loop.

File: src/audio/copier/copier.c
Comment (btian1) [reply src/audio/copier/copier.c:728]: yes, I will change to this style, previously focused on resolve multisink and other CI related errors.

File: src/audio/dai-zephyr.c
Comment (btian1) [reply src/audio/dai-zephyr.c:381]: ok

File: src/audio/copier/copier.c
Comment (btian1) [reply src/audio/copier/copier.c:889]: switch will cover 3 cases, host/dai/gtw, anything not covered?

File: src/audio/dai-zephyr.c
Comment (btian1) [reply src/audio/dai-zephyr.c]: keep it for now, I tried with remove this part, I saw some input/output error with capture:
arecord: pcm_read:2221: read error: Input/output error

File: src/audio/dai-zephyr.c
Comment (btian1) [reply src/audio/dai-zephyr.c:1053]: done

File: src/audio/copier/copier.c
Comment (btian1) [reply src/audio/copier/copier.c:1781]: already merged PR: #7303

File: src/audio/copier/copier.c
Comment (btian1) [reply src/audio/copier/copier.c:1797]: done

File: src/audio/copier/copier.c
Comment (btian1) [reply src/audio/copier/copier.c:1817]: ok, changed aligned back.

File: src/audio/copier/copier.c
Comment (btian1) [reply src/audio/copier/copier.c:1819]: removed

File: src/audio/copier/copier.c
Comment (btian1) [reply src/audio/copier/copier.c:1861]: done

File: src/ipc/ipc4/dai.c
Comment (btian1) [reply src/ipc/ipc4/dai.c]: agree, just because this is an formal inerface for driver, if move ipc4 to inline, then legacy part also need inline, it is not a performance function, so overhead is not a problem.

File: src/audio/copier/copier.c
Comment (btian1) [reply src/audio/copier/copier.c:1543]: added

File: src/audio/dai-zephyr.c
Comment (btian1) [reply src/audio/dai-zephyr.c:1358]: it is part of drv interface, and copier actually does not use dai_copy, it is using dai_zephyr_copy.

File: src/audio/copier/copier.c
Comment (ranj063) [inline src/audio/copier/copier.c:1024]: ahh i forgot a module copier is not of type DAI or HOST. Please remove all errors as there's just nothing to do for that case.

File: src/audio/copier/copier.c
Comment (ranj063) [inline src/audio/copier/copier.c]: why not dev instead of cd->endpoint[0]? and why remove the loop?

**Review by ranj063 — CHANGES_REQUESTED** [review]
I'm going to stop at this commit @btian1. I think you need to fix the multiple endpoint case right from the begininng and also add a common header

File: src/audio/copier/copier.c
Comment (btian1) [reply src/audio/copier/copier.c:739]: yes, this part should not be error, I observed NONE with endpoint_num = 0 case, left a warning is more suitable.

File: src/audio/dai-legacy.c
Comment (btian1) [reply src/audio/dai-legacy.c:236]: I also don't want to touch, but I remember dai is different, if not touch, I can't make CI pass.

File: src/audio/dai-zephyr.c
Comment (btian1) [reply src/audio/dai-zephyr.c:396]: ok, like new/free right? if dma_get failed, need dma_put.

File: src/include/sof/lib/dai-zephyr.h
Comment (btian1) [reply src/include/sof/lib/dai-zephyr.h:295]: there is a lot of things need do, so my thinking is: first finish this PR's mission, then use other PR to do other things, I have to change legacy part, I remember, otherwise, CI build will report error.

File: src/include/sof/audio/dai_copier.h
Comment (btian1) [reply src/include/sof/audio/dai_copier.h:19]: do you mean delete those comments? please confirm, if so, I will drop.

File: src/audio/dai-zephyr.c
Comment (btian1) [reply src/audio/dai-zephyr.c:447]: should change or delete, it is a self-remind comments when I worked on it.

File: src/audio/copier/copier.c
Comment (btian1) [reply src/audio/copier/copier.c:904]: I think switch is more clear, also Guennadi request this change, another advantage is better for expand, just add new case, if use "if else", logic will make reviewer crazy.

File: src/audio/dai-legacy.c
Comment (btian1) [reply src/audio/dai-legacy.c:621]: this can be changed, dd is newly added.

File: src/audio/dai-zephyr.c
Comment (btian1) [reply src/audio/dai-zephyr.c:985]: ok, I previously think ret maybe 0, may not be suitable when xrun happens.

File: src/include/sof/lib/dai-legacy.h
Comment (btian1) [reply src/include/sof/lib/dai-legacy.h:546]: ok, just need more changes.

File: src/audio/copier/copier.c
Comment (btian1) [reply src/audio/copier/copier.c:938]: endpoint_num always 1, right? is there a  case for endpoint_num >= 2?

File: src/audio/copier/copier.c
Comment (btian1) [reply src/audio/copier/copier.c]: this is just change to zephyr reset, for replace, I made it in patch 11 together.

File: src/include/sof/lib/dai-legacy.h
Comment (btian1) [reply src/include/sof/lib/dai-legacy.h:551]: ok

File: src/audio/copier/copier.c
Comment (btian1) [reply src/audio/copier/copier.c:1024]: ok, remove error is fine, for moduler copier, what's the assigned type, or it does not need a type?

File: src/audio/copier/copier.c
Comment (btian1) [reply src/audio/copier/copier.c]: change dev with patch 11.

File: src/audio/copier/copier.c
Comment (ranj063) [reply src/audio/copier/copier.c:904]: remove the error in the default case as this will be the module copier and we dont need to do anything for that

Comment (btian1) [reply src/audio/copier/copier.c:904]: sure

File: src/audio/dai-zephyr.c
Comment (lyakh) [reply src/audio/dai-zephyr.c:396]: yes, like new / free, that's why - no, not `dma_put()`, but `dai_put()`, because `dai_get()` was successful above

File: src/audio/dai-legacy.c
Comment (btian1) [reply src/audio/dai-legacy.c:236]: will do it with a separate patch after this PR merged.

Comment (lyakh) [reply src/audio/dai-legacy.c:236]: if DAI merge with copier is any similar to what previously was done to host, then we do need to touch legacy: https://github.com/thesofproject/sof/pull/7443/commits/69deb6f112dc3336d527f5e34f49d927bbbac5ab

File: src/audio/copier/copier.c
Comment (serhiy-katsyuba-intel) [reply src/audio/copier/copier.c:938]: Yes, there is a case when endpoint_num is >= 2. It's called "multi-channel ALH" or "ALH multi-gateway" feature. Basically, there could be multiple ALH endpoints audio data from/to which should be multiplexed/demultiplexed into/from single stream. So, for example, stereo audio could come from two ALH gateways, one for left channel, second for right, and then that audio is multiplexed into single stereo stream. See mux_into_multi_endpoint_buffer() and demux_from_multi_endpoint_buffer().

**Comment by serhiy-katsyuba-intel** [comment]
@btian1 , before merge could you ask @wszypelt to run a full scope of Windows MTL tests on firmware with this PR. Unfortunately, CI is only done on some subset of tests. After similar Host copy optimizations we had some glitches/distorted audio regressions. Full scope tests take several hours, I guess, that is why they are not all added to CI.

I'll put "request changes" just to block merge until @wszypelt is back with the tests results.

**Review by serhiy-katsyuba-intel — CHANGES_REQUESTED** [review]
Blocking merge until @wszypelt gets back with full scope tests results.

**Comment by wszypelt** [comment]
@serhiy-katsyuba-intel @btian1 Sure thing, I can set up full scope, within 24 hours we should have test results for quick comparison
For now PR also blocks CI(Internal Intel CI System/merge/build— Tests failed ), unfortunately the sdw tests don't pass properly

**Comment by btian1** [comment]
> unfortuna

please stay tuned, will continue on ww17, need more time to work on this PR.

**Pull request closed without merging**
